src/exo/range_analysis.py

```diff
@@ -1,104 +1,171 @@
+from __future__ import annotations
 from collections import ChainMap
+from dataclasses import dataclass
+from typing import Optional
 
-from .LoopIR import LoopIR, T
+from .LoopIR import LoopIR, T, LoopIR_Compare, get_reads_of_expr
 from .new_eff import Check_ExprBound, Check_ExprBound_Options
 
 
-def index_range_analysis(expr, env):
+def binop(op: str, e1, e2):
+    return LoopIR.BinOp(op, e1, e2, e1.type, e1.srcinfo)
+
+
+@dataclass
+class IndexRange:
     """
-    Performs range-analysis on an index expression.
-
-    Function takes in an environment in which the expression exists.
-    The environment is a mapping from `Sym` -> range.
-    It calculates the range of possible values that the index
-    expression could be in.
-
-    Any range, in the environment mapping or the result of the
-    analysis is either:
-        A tuple `T` of length 2 s.t. it represents the range
-        `[T[0], T[1]]` (both inclusive).
-        If `T[0]` or `T[1]` is `None` it represents no knowledge
-        of the value of that side of the range.
+    Represents a range of possible values between [base + lo, base + hi]
+        - [base] contains all expressions that we don't want to bounds infer (e.g. free
+        variables). If base is None, it signifies a base of 0.
```

Comment:
I think representing the base = None as LoopIR.Const(0) might simplify a lot of the code later.

Reply:
The problem with this approach is that we would either (a) generate a lot of 0s from all the interval merging and yield expressions like `0 * 0 + 0 + i` or (b) effectively have to check against `LoopIR.Const(0, blah)` to avoid the issue in (a), which is essentially what I'm already doing with `None` now anyways.